src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,687 @@
+// vim: ts=8 sw=2 smarttab
+
+#include <string.h>
+
+#include "include/buffer.h"
+#include "include/byteorder.h"
+#include "crimson/os/seastore/transaction_manager.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "seastar/core/thread.hh"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::omap_manager {
+
+std::ostream &operator<<(std::ostream &out, const omap_inner_key_t &rhs)
+{
+  return out << "omap_inner_key (" << rhs.key_off<< " - " << rhs.key_len
+             << " - " << rhs.laddr << ")";
+}
+
+std::ostream &operator<<(std::ostream &out, const omap_leaf_key_t &rhs)
+{
+  return out << "omap_leaf_key_t (" << rhs.key_off<< " - " << rhs.key_len
+             << " "<< rhs.val_off<<" - " << rhs.val_len << ")";
+}
+
+std::ostream &OMapInnerNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << get_meta().depth;
+}
+
+OMapInnerNode::handle_splitted_ret
+OMapInnerNode::handle_splitted(omap_context_t oc, internal_iterator_t iter,
+                               mutation_result_t mresult)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  auto [left, right, pivot] = *(mresult.split_tuple);
+  if (!extent_will_overflow(pivot.size() + 1, std::nullopt)) {
+    if (!is_pending()) {
+      auto mut = oc.tm.get_mutable_extent(oc.t, this)->cast<OMapInnerNode>();
+      auto mut_iter = mut->iter_idx(iter.get_index());
+      return mut->handle_splitted(oc, mut_iter, mresult);
+    }
+    journal_inner_update(iter, left->get_laddr(), maybe_get_delta_buffer());
+    journal_inner_insert(iter + 1, right->get_laddr(), pivot,
+                         maybe_get_delta_buffer());
+    return insert_ret(
+           insert_ertr::ready_future_marker{},
+           mutation_result_t(mutation_status_t::SUCCESS, std::nullopt, std::nullopt));
+  } else {
+    return make_split_children(oc)
+      .safe_then([this, oc, iter, mresult = std::move(mresult)] (auto tuple) {
+       OMapNodeRef pleft = std::get<0>(tuple);
+       OMapNodeRef pright = std::get<1>(tuple);
+       std::string p_pivot = std::get<2>(tuple);
+       OMapInnerNodeRef p_left = pleft->cast<OMapInnerNode>();
+       OMapInnerNodeRef p_right = pright->cast<OMapInnerNode>();
+       auto [c_left, c_right, c_pivot] = *(mresult.split_tuple);
+       if (c_pivot < p_pivot) {  //left
+         auto mut_iter = p_left->iter_idx(iter.get_index());
+         p_left->journal_inner_update(mut_iter, c_left->get_laddr(),
+                                      p_left->maybe_get_delta_buffer());
+         p_left->journal_inner_insert(mut_iter + 1, c_right->get_laddr(), c_pivot,
+                                      p_left->maybe_get_delta_buffer());
+       } else if(c_pivot == p_pivot) {
```

Comment:
It this case actually possible?  p_pivot came from splitting the existing *this and must therefore be one of the keys in *this.  c_pivot necessarily falls within a node from *this (and not the first key) so it must not be equal to p_pivot.  Therefore, I think the two new nodes must both be inserted either into p_left or p_right -- they can't be split.